### DISCUSSION TITLE
How to make handleSubmit play nice with TypeScript?

### DISCUSSION THREAD

**Discussion by ifeltsweet** [react-hook-form/react-hook-form#6195] [Q&A]
My main gripe with the way built-in validations work is that the validation rules have no effect on the type provided to `handleSubmit`. Consider the following example:

```tsx
type FormInputs = {
  name: string | null;
}

function Form() {
  const { register, handleSubmit } = useForm<FormInputs>({
    defaultValues: {
      name: null,
    }
  });

  const onSubmit = handleSubmit((data) => {
    data.name // the type is string | null but should be string.

    // Workaround 1.
    data.name! // the type is string.

    // Workaround 2.
    if (!data.name) return;
    data.name // the type is string.
  });

  return (
    <form onSubmit={onSubmit}>
      <input {...register("name", { required: true })} />
      <input type="submit" />
    </form>
  );
}
```

Apart from the two workarounds shown above, I can only think of not using built-in validation and instead use something like zod and alias data in the handleSubmit to a validated form type that zod would provide.

However, I think this could be partially solved if rules for validations are defined in the useForm hook so that a valid form type can be inferred from them:

```ts
useForm({ rules: { name: { required: true } } })
```

Obviously, this is not going to help with custom validators and is really only useful for the `required` validator.

How are you dealing with this issue?

**Comment by jorisre** [comment]
Is it possible to change to:

```ts
type FormInputs = {
  name: string;
}
```

And

```ts
const { register, handleSubmit } = useForm<FormInputs>({
    defaultValues: {
      name: '', // or undefined ?
    }
});
```

**Reply by ifeltsweet** [reply]
In this simple example, it is possible, but not so much when you have a non-string property.

```ts
type FormInputs = {
  age: number | null; // We don't want default to be 0, we want it not set at all.
  gender: 'male' | 'famale' | null // We don't want to preselect any specific gender as default.
  appointmentDate: Date | null // We don't want to preselect any specific date.
  location: Location | null; // We don't want to preselect any specific location. Location is an object.
}
```
